File: eduafrica-backend/eduafrica-backend/src/main/java/com/eduafrica/service/FormationService.java
package com.eduafrica.service;

import com.eduafrica.dto.FormateurStatsResponse;
import com.eduafrica.model.Formation;
import com.eduafrica.model.User;
import com.eduafrica.model.enums.FormationCategory;
import com.eduafrica.model.enums.FormationLevel;
import com.eduafrica.repository.FormationRepository;
import com.eduafrica.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.util.List;

@Service
public class FormationService {
    
    @Autowired
    private FormationRepository formationRepository;
    
    @Autowired
    private UserRepository userRepository;
    
    public Formation createFormation(Formation formation, String formateurEmail) {
        User formateur = userRepository.findByEmail(formateurEmail)
                .orElseThrow(() -> new RuntimeException("Formateur non trouvé"));
        
        formation.setFormateur(formateur);
        return formationRepository.save(formation);
    }
    
    public Formation updateFormation(Long id, Formation formationDetails) {
        Formation formation = formationRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Formation non trouvée"));
        
        formation.setTitle(formationDetails.getTitle());
        formation.setDescription(formationDetails.getDescription());
        formation.setProgramme(formationDetails.getProgramme());
        formation.setCategory(formationDetails.getCategory());
        formation.setLevel(formationDetails.getLevel());
        formation.setPrice(formationDetails.getPrice());
        formation.setDuration(formationDetails.getDuration());
        formation.setTags(formationDetails.getTags());
        formation.setImageUrl(formationDetails.getImageUrl());
        
        return formationRepository.save(formation);
    }
    
    public void deleteFormation(Long id) {
        formationRepository.deleteById(id);
    }
    
    public Formation getFormationById(Long id) {
        return formationRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Formation non trouvée"));
    }
    
    public Page<Formation> getAllFormations(int page, int size) {
        Pageable pageable = PageRequest.of(page, size, Sort.by("createdAt").descending());
        // Utiliser la méthode avec JOIN FETCH pour charger le formateur
        Page<Formation> formations = formationRepository.findAllWithFormateur(pageable);
        return formations;
    }
    
    public Page<Formation> searchFormations(String keyword, int page, int size) {
        Pageable pageable = PageRequest.of(page, size);
        return formationRepository.searchByKeyword(keyword, pageable);
    }
    
    public Page<Formation> getFormationsByFilters(
            FormationCategory category,
            FormationLevel level,
            Boolean isFree,
            int page,
            int size
    ) {
        Pageable pageable = PageRequest.of(page, size);
        return formationRepository.findByFilters(category, level, isFree, pageable);
    }
    
    public List<Formation> getFormationsByFormateur(String formateurEmail) {
        User formateur = userRepository.findByEmail(formateurEmail)
                .orElseThrow(() -> new RuntimeException("Formateur non trouvé"));
        return formationRepository.findByFormateur(formateur);
    }
    
    public FormateurStatsResponse getFormateurStats(String formateurEmail) {
        User formateur = userRepository.findByEmail(formateurEmail)
                .orElseThrow(() -> new RuntimeException("Formateur non trouvé"));
        List<Formation> formations = formationRepository.findByFormateur(formateur);
        
        long totalFormations = formations.size();
        long totalStudents = formations.stream()
                .mapToLong(f -> f.getNbStudents() != null ? f.getNbStudents() : 0)
                .sum();
        
        double averageRating = formations.stream()
                .filter(f -> f.getAverageRating() != null && f.getAverageRating() > 0)
                .mapToDouble(Formation::getAverageRating)
                .average()
                .orElse(0.0);
        
        BigDecimal estimatedRevenue = formations.stream()
                .filter(f -> !f.getIsFree() && f.getPrice() != null && f.getNbStudents() != null)
                .map(f -> f.getPrice().multiply(BigDecimal.valueOf(f.getNbStudents())))
                .reduce(BigDecimal.ZERO, BigDecimal::add);
        
        return FormateurStatsResponse.builder()
                .totalFormations(totalFormations)
                .totalStudents(totalStudents)
                .averageRating(averageRating)
                .estimatedRevenue(estimatedRevenue)
                .build();
    }
}

File: eduafrica-frontend/eduafrica-frontend/src/app/core/services/formation.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Formation, FormationFilters, PageResponse } from '../../shared/models/formation.model';

@Injectable({
  providedIn: 'root'
})
export class FormationService {
  private apiUrl = 'http://localhost:8080/api/formations';

  constructor(private http: HttpClient) {}

  getAllFormations(page: number = 0, size: number = 12): Observable<PageResponse<Formation>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http.get<PageResponse<Formation>>(this.apiUrl, { params });
  }

  getFormationById(id: number): Observable<Formation> {
    return this.http.get<Formation>(`${this.apiUrl}/${id}`);
  }

  searchFormations(keyword: string, page: number = 0, size: number = 12): Observable<PageResponse<Formation>> {
    const params = new HttpParams()
      .set('keyword', keyword)
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http.get<PageResponse<Formation>>(`${this.apiUrl}/search`, { params });
  }

  filterFormations(filters: FormationFilters, page: number = 0, size: number = 12): Observable<PageResponse<Formation>> {
    let params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());

    if (filters.category) {
      params = params.set('category', filters.category);
    }
    if (filters.level) {
      params = params.set('level', filters.level);
    }
    if (filters.isFree !== undefined) {
      params = params.set('isFree', filters.isFree.toString());
    }

    return this.http.get<PageResponse<Formation>>(`${this.apiUrl}/filter`, { params });
  }

  createFormation(formation: Partial<Formation>): Observable<Formation> {
    return this.http.post<Formation>(this.apiUrl, formation);
  }

  updateFormation(id: number, formation: Partial<Formation>): Observable<Formation> {
    return this.http.put<Formation>(`${this.apiUrl}/${id}`, formation);
  }

  deleteFormation(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  getMyFormations(): Observable<Formation[]> {
    return this.http.get<Formation[]>(`${this.apiUrl}/my-formations`);
  }

  getFormateurStats(): Observable<{totalFormations: number, totalStudents: number, averageRating: number, estimatedRevenue: number}> {
    return this.http.get<{totalFormations: number, totalStudents: number, averageRating: number, estimatedRevenue: number}>(`${this.apiUrl}/formateur/stats`);
  }
}
